notify/templates.go

```diff
@@ -0,0 +1,240 @@
+)
+
+type TestTemplatesConfigBodyParams struct {
+	// Alerts to use as data when testing the template.
+	Alerts []*PostableAlert `json:"alerts"`
+
+	// Template string to test.
+	Template string `json:"template"`
+
+	// Name of the template file.
+	Name string `json:"name"`
+}
+
+type TestTemplatesResults struct {
+	Results []TestTemplatesResult      `json:"results"`
+	Errors  []TestTemplatesErrorResult `json:"errors,omitempty"`
+}
+
+type TestTemplatesResult struct {
+	// Name of the associated template definition for this result.
+	Name string `json:"name"`
+
+	// Interpolated value of the template.
+	Text string `json:"text"`
+}
+
+type TestTemplatesErrorResult struct {
+	// Name of the associated template for this error. Will be empty if the Kind is "invalid_template".
+	Name string `json:"name,omitempty"`
+
+	// Kind of template error that occurred.
+	Kind TemplateErrorKind `json:"kind"`
+
+	// Error message.
+	Message string `json:"message"`
+}
+
+type TemplateErrorKind string
+
+const (
+	InvalidTemplate TemplateErrorKind = "invalid_template"
+	ExecutionError  TemplateErrorKind = "execution_error"
+)
+
+// TestTemplate tests the given template string against the given alerts. Existing templates are used to provide context for the test.
+// If an existing template of the same filename as the one being tested is found, it will not be used as context.
+func (am *GrafanaAlertmanager) TestTemplate(ctx context.Context, c TestTemplatesConfigBodyParams) (*TestTemplatesResults, error) {
+	definitions, err := parseTestTemplate(c.Name, c.Template)
+	if err != nil {
+		return &TestTemplatesResults{
+			Errors: []TestTemplatesErrorResult{{
+				Kind:    InvalidTemplate,
+				Message: err.Error(),
+			}},
+		}, nil
+	}
+
+	// Recreate the current template without the definition blocks that are being tested. This is so that any blocks that were removed don't get defined.
+	paths := make([]string, 0)
+	for _, name := range am.templates {
+		if name == c.Name {
+			// Skip the existing template of the same name as we're going to parse the one for testing instead.
+			continue
+		}
+		paths = append(paths, filepath.Join(am.workingDirectory, name))
+	}
+
+	// Parse current templates.
+	var newTextTmpl *tmpltext.Template
+	var captureTemplate template.Option = func(text *tmpltext.Template, _ *tmplhtml.Template) {
+		newTextTmpl = text
+	}
+	newTmpl, err := templates.FromGlobs(paths, captureTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse test template.
+	_, err = newTextTmpl.Parse(c.Template)
+	if err != nil {
+		// This shouldn't happen since we already parsed the template above.
+		return nil, err
+	}
+
+	externalURL, err := url.Parse(am.ExternalURL())
+	if err != nil {
+		return nil, err
+	}
+	newTmpl.ExternalURL = externalURL
+
+	// Prepare the context.
+	alerts := v2.OpenAPIAlertsToAlerts(c.Alerts)
+	ctx = notify.WithReceiverName(ctx, "test receiver")
+	ctx = notify.WithGroupLabels(ctx, model.LabelSet{"group_label": "group_label_value"})
```

Comment:
why do we need that? If we do need that then probably we should group c.Alerts and provide a correct label set?

Reply:
`GetTemplateData` will error if they don't exist. Otherwise, they are available as template values. We could group technically group them, but what would we use as the `groupby`? Do we actually use their notification policies or use a default of `folder, alertname`?

Reply:
Good question, I do not know. Probably as an alternative, we can group by common labels, or let user choose (in the API request)

Reply:
I'm inclined to leave this as-is for now, it's not clear how to go about this. We can improve it in the next iteration.